File: cpp/questions/lengthOfLIS.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int ns = nums.size();
        int ans = 0;
        std::vector<int> ep(ns, 0);
        for (const auto n: nums) {
            int l = 0;
            int r = ans;
            while (l < r) {
                int m = (l + r) / 2;
                if (ep[m] >= n) {
                    r = m;
                } else {
                    l = m + 1;
                }
            }
            ep[l] = n;
            ans = std::max(ans, l+1);
        }
        return ans;        
    }
};

File: python/questions/lengthOfLIS.py
from typing import List

class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        ln = len(nums)
        ans = 0
        ep = [0] * ln
        for n in nums:
            l = 0
            r = ans
            while l < r:
                m = (l + r) // 2
                if ep[m] >= n:
                    r = m
                else:
                    l = m + 1
            ep[l] = n
            ans = max(ans, l+1)
        return ans
